src/inmanta/deploy/scheduler.py

```diff
@@ -641,7 +660,7 @@ async def report_resource_state(
         async with self._scheduler_lock:
             # refresh resource details for latest model state
             details: Optional[ResourceDetails] = self._state.resources.get(resource, None)
-            if details is None or details.attribute_hash != attribute_hash:
+            if details is None or details.status is ResourceStatus.UNDEFINED or details.attribute_hash != attribute_hash:
```

Comment:
I think has to do with an issue I encountered in the tests. In the tests it's possible to have a resource in the undefined state that doesn't have an `<<undefined>>` attribute. So you can have a resource A of version X that is undefined, while the same resource in version X+1 is not undefined and has the same resource hash. Conclusion: I think you are right that the undefined check can be removed here, because the above-mentioned scenario cannot happen in practice.

Reply:
I see. I'm also not against keeping it and adding a comment that the scenario can happen in test cases. I'm typically against tailoring the code for the tests, but in this case the check is redundant, but still conceptually correct, so I'm more lenient.